### ISSUE TITLE
The prediction result is not correct

### ISSUE THREAD

**Issue by vlzzhou** [spthermo/generative-image-inpainting#3]
![image](https://user-images.githubusercontent.com/38495504/120413140-568e4680-c38a-11eb-957d-df8fbfbe60a3.png)
The result is strange. Do you know what caused this situation?

**Comment by iamjak13** [comment]
I guess this is a snapshot of the generator's output. From the snapshot I can see that the values are in the correct range (since we use tanh as activation the values are expected between -1 and 1). Why do you think that the result is strange? Did you visualize the predicted image?

**Comment by spthermo** [comment]
I guess this is a snapshot of the generator's output. From the snapshot I can see that the values are in the correct range (since we use tanh as activation the values are expected between -1 and 1). Why do you think that the result is strange? Did you visualize the predicted image?

**Comment by vlzzhou** [comment]
I visualized the resulting image, which is a black image
![image](https://user-images.githubusercontent.com/38495504/121311851-63390e80-c937-11eb-8b23-f7ca26ee3556.png)
![image](https://user-images.githubusercontent.com/38495504/121311885-6cc27680-c937-11eb-8014-dab570ebb250.png)
How can the visualization be correct?

**Comment by vlzzhou** [comment]
> I guess this is a snapshot of the generator's output. From the snapshot I can see that the values are in the correct range (since we use tanh as activation the values are expected between -1 and 1). Why do you think that the result is strange? Did you visualize the predicted image?

Hello, I visualized the forecast result in this way, but the result is still wrong. What should I do? Very gratefully
            inpainted_image = generated_samples[j]
            img = inpainted_image.clone().add(1).div(2).mul(255).clamp(0, 255).cpu().detach().numpy()
            img = img.transpose(1, 2, 0).astype("uint8")
            img = Image.fromarray(img)
            img.save('Inpainted_Output.png')

![image](https://user-images.githubusercontent.com/38495504/121641030-40d2fc80-cac1-11eb-8f09-0dd3623ebc22.png)

**Comment by spthermo** [comment]
In the inference.py script there is a line (161) that saves the generated images locally (at a given path):

`
scipy.misc.toimage(generated_samples.view(-1, 3, 128, 128)[0].cpu().detach().numpy()).save(out_path)
`

Did you try this code? 

Having an null image is weird, can you put a breakpoint at the first line of the code snippet you shared to see the values of generated_samples[j]? e.g. mean/std, min, max

**Comment by vlzzhou** [comment]
> In the inference.py script there is a line (161) that saves the generated images locally (at a given path):
> 
> `scipy.misc.toimage(generated_samples.view(-1, 3, 128, 128)[0].cpu().detach().numpy()).save(out_path)`
> 
> Did you try this code?
> 
> Having an null image is weird, can you put a breakpoint at the first line of the code snippet you shared to see the values of generated_samples[j]? e.g. mean/std, min, max

This code cannot be used due to version issues. My python version is .3.8, and there is no lower version of scipy that matches it.
![image](https://user-images.githubusercontent.com/38495504/122146096-008cc900-ce89-11eb-850d-a7ab7a2f4a6d.png)
![image](https://user-images.githubusercontent.com/38495504/122146122-0bdff480-ce89-11eb-8ec4-d02543a7f8c9.png)
How can I solve this problem?

**Comment by spthermo** [comment]
Hi again,

use this save image code snippet: [code](https://github.com/VCL3D/DeepDepthDenoising/blob/master/exporters/image.py)
it requires opencv-python. I've just run a quick example saving the images of the final batch every epoch locally and it works fine.

Use the following snippet to save:
`
save_image('sample#.jpg', gen_data.view(-1, 3, opt.imageSize, opt.imageSize))
`

I hope this helps.

**Comment by vlzzhou** [comment]
> Hi again,
> 
> use this save image code snippet: [code](https://github.com/VCL3D/DeepDepthDenoising/blob/master/exporters/image.py)
> it requires opencv-python. I've just run a quick example saving the images of the final batch every epoch locally and it works fine.
> 
> Use the following snippet to save:
> `save_image('sample#.jpg', gen_data.view(-1, 3, opt.imageSize, opt.imageSize))`
> 
> I hope this helps.

Thank you very much for your prompt reply. But I failed again, please see this is the data before saving. I suspect that there is a problem with the model I trained. Could you please upload your model data?
![image](https://user-images.githubusercontent.com/38495504/122321893-2e404380-cf57-11eb-8b20-3d4515a37b21.png)
![image](https://user-images.githubusercontent.com/38495504/122321977-4adc7b80-cf57-11eb-8aa6-0edf86e29d81.png)
And the save-image function just takes the picture out and multiplies it by 255.0 to enlarge it. The data before this is tanh. According to the discussion with you before, it is also between (-1,1). This is simply too weird.

**Comment by spthermo** [comment]
You are right about the normalization, so you can just normalize the array to [0,1] and then apply the scale, e.g.:
`
array = scale * (array.transpose(1, 2, 0) - numpy.min(array)) / (numpy.max(array) - numpy.min(array))
`
running the model really quickly for some CelebA images leads to the following snapshots:
![1](https://user-images.githubusercontent.com/1241918/122368842-9726bc00-cf55-11eb-9c7d-6b18042455be.jpg)

and moving to the next line:
![2](https://user-images.githubusercontent.com/1241918/122368904-a574d800-cf55-11eb-8d7c-29bb7b33f70d.jpg)

Other than that, if your input data are normalized between [-1,1], I don't see any reason for any training problems related to the model. Right now I don't have a ready to release model (it's kind of an "old" repo), but I will try to run something fast and release it. However, I strongly recommend you to keep working on your code as the model has been thoroughly tested. Check your input, the hyper-parameters used during training, number of epochs, etc.
